runtime/wasm/src/host.rs

```diff
@@ -684,103 +689,177 @@ impl RuntimeHostTrait for RuntimeHost {
-                        format_err!(
-                            "Mapping terminated before finishing to handle \
-                             Ethereum event: {}",
-                            event_signature,
-                        )
+        // Filter out handlers whose corresponding event ABIs cannot decode the
+        // params (this is common for overloaded events that have the same topic0
+        // but have indexed vs. non-indexed params that are encoded differently).
+        //
+        // Map (handler, event ABI) pairs to (handler, decoded params) pairs.
+        let matching_handlers = valid_handlers
+            .into_iter()
+            .filter_map(|(event_handler, event_abi)| {
+                event_abi
+                    .parse_log(RawLog {
+                        topics: log.topics.clone(),
+                        data: log.data.clone().0,
                     })
-                })
-                .and_then(move |(result, send_time)| {
-                    info!(
-                        logger, "Done processing Ethereum event";
-                        "signature" => &event_handler.event,
-                        "handler" => &event_handler.handler,
-                        "total_ms" => start_time.elapsed().as_millis(),
-
-                        // How much time the result spent in the channel,
-                        // waiting in the tokio threadpool queue. Anything
-                        // larger than 0 is bad here. The `.wait()` is instant.
-                        "waiting_ms" => send_time.wait().unwrap().elapsed().as_millis(),
-                    );
-                    result
-                }),
-        )
+                    .map(|log| log.params)
+                    .map_err(|e| {
+                        info!(
+                            logger,
+                            "Skipping handler because the event parameters do not \
+                             match the event signature. This is typically the case \
+                             when parameters are indexed in the event but not in the \
+                             signature or the other way around";
+                            "handler" => &event_handler.handler,
+                            "event" => &event_handler.event,
+                            "error" => format!("{}", e),
+                        );
+                    })
+                    .ok()
+                    .map(|params| (event_handler, params))
+            })
+            .collect::<Vec<_>>();
+
+        // If none of the handlers match the event params, fail processing the event
+        if matching_handlers.is_empty() {
+            return Box::new(future::err(format_err!(
+                "No matching event handlers found for in data source \"{}\"",
+                data_source_name,
+            )));
+        }
+
+        // Process the event with all matching handlers; typically there is only one
```

Comment:
An event can trigger more than one handler? In what case does this happen? Is it something we want to allow at this point, are we confident that they're processed in a deterministic order?

IMO, until a use case actually requires something more flexible, we should keep a 1:1 relationship between triggers and handlers for the sake of simplicity, though it doesn't look like we actually validate that atm.

Reply:
We're confident about the deterministic order, because the handlers are matched and executed in the same order they appear in the manifest.

We should disallow it in the manifest; I've filed https://github.com/graphprotocol/graph-cli/issues/286 to track that. Will update the code here to just use the first match.